src/local.rs

```diff
@@ -105,29 +92,45 @@ impl Local {
         )
         .context(OpenDatabaseSnafu {})?;
 
-        // Get the relative directory of the maildir to the database path.
-        let canonical_db_path = db.path().canonicalize().context(CanonicalizeSnafu {})?;
-        let canonical_mail_dir_path = mail_dir
-            .as_ref()
+        // Get notmuch's idea of the mail root. If, for whatever reason, we get nothing back for
+        // that key (ancient version of notmuch?), use the database path.
+        let mail_root = db
+            .config(ConfigKey::MailRoot)
+            .map_or(
+                db.path().into(),
+                |root| PathBuf::from(root)
+            )
             .canonicalize()
             .context(CanonicalizeSnafu {})?;
-        let relative_mail_dir = canonical_mail_dir_path
-            .strip_prefix(&canonical_db_path)
-            .context(MailDirNotASubdirOfNotmuchRootSnafu {
-                mail_dir: &canonical_mail_dir_path,
-                notmuch_root: &canonical_db_path,
-            })?;
 
-        // Build the query to search for all mail in our maildir.
-        let all_mail_query = format!("path:\"{}/**\"", relative_mail_dir.to_str().unwrap());
+        // Figure out our maildir. Either the configured thing, or notmuch's mail root. Which in
+        // the worst case will be notmuch's database dir, but that's probably not the worst choice.
+        let mail_dir = match &config.mail_dir {
+            Some(ref dir) => dir.clone(),
+            _             => mail_root.clone(),
+        }.canonicalize().context(CanonicalizeSnafu {})?;
+        debug!("mail dir: {}", mail_dir.to_string_lossy());
+
+        // Build the query to search for all mail in our maildir. If the maildir is under the
+        // notmuch mail root, then search under the relative maildir path (allowing multiple
+        // maildirs per notmuch dir). If not, assume this is the only maildir for the notmuch dir,
+        // and use a global query.
+        let all_mail_query = mail_dir
```

Comment:
If the user types an absolute path for the `mail_dir` config option but misspells part of the path, causing `strip_prefix` to fail, they could experience some nasty consequences. We should keep/revise the `MailDirNotASubdirOfNotmuchRoot` error here in that case instead of swallowing. We can still support the config option being a relative path by checking if its `is_absolute()` returns false and handling that slightly differently.

Reply:
I'm not sure there's a lot we can do really; how do we know if its a  error if the path is valid?

The best I can think to do is for a configured `mail_dir`, make sure it exists and bail out if not. Of course its still possible to typo, but at least it won't plow on regardless.

Reply:
Ok, I've added a commit that does just that if `mail_dir` is provided (if we compute it, it doesn't have to exist beforehand).

The more I think about the more I think this will be rare, because most of the time I expect that people would mess with their config until its set up how they like, and would then walk away, so its unlikely that they're going to get a path wrong in a system they've already set up. Its still good to have a safety for an obviously-wrong situation though.

Reply:
Hmm, this isn't quite what I meant. I think I understand why you changed it like this now, to specially handle the case for if `mail_dir` and the notmuch root directory are the same, is that right? According to the docs, it seems like `strip_prefix` will yield an empty `PathBuf` rather than `Err` for this case, so there's actually no need to specially handle this case.

https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.strip_prefix

But the consequences of removing the `MailDirNotASubdirOfNotmuchRoot` error as this commit currently exists would result in something like the following:

- The notmuch root directory is `~/Maildir` and mujmap is configured so that its mail dir is `~/Maildir/mujmap`. So far so good.
- The user decides they want to move their maildir to an external hard drive, and move it to `/mnt/maildrive/Maildir`, but forget to update mujmap, which still points to `~/Maildir`
- The user runs mujmap. Both directories are valid, but because `strip_prefix` fails here, mujmap silently both decides that it owns the entire mail database, and insists on placing mail files in the old `~/Maildir` and trying to add them to the notmuch database. Potentially destructive chaos ensues.

Reply:
I'm not really following your scenario, and I'm not sure if that's because its imprecisely described, or if its because we're not quite in agreement about what we're doing here and so talking past a little bit. I will try and write down what I think is happening here, and hopefully I will understand by the end.

I'm seeing `mail_dir` (and `state_dir` and `cache_dir`) as a kind of override; if you set it, it _will_ be used over any other value.

So in your scenario, if they've set `mail_dir` to something `~/Maildir/mujmap` then they get what they ask for, regardless of whether or not its a subdir of the notmuch root. This has to be this way to support "side-by-side" configs (like XDG).

If on the other hand they haven't set `mail_dir` explicitly and its implied (by "directory" mode), then I don't see how mujmap could know this. The paths haven't changed, just the files were moved out without its knowledge.

Going back to the code, what I think I'm doing is that if the maildir is under the notmuch root, then the query becomes `path:xxx/**` where `xxx` is the path of the maildir relative to the notmuch root, but if its outside the notmuch root, we use the global `path:**` because what else can we do? The code is effectively converting the empty `PathBuf` and `Err` to mean the same thing: use a global query.

Ahh, maybe that's it. Are you talking about the case where the user has explicitly set `mail_dir` to be under the notmuch root? And then they move it later, which will change the path prefix. That could be a problem (I have not thought hard about this yet). I hadn't considered that someone could deliberately set things up this way because I only know of yours and my use cases! Hmm.

I wonder if that's the real problem here, that the path prefix is kind of baking too much info about the existing paths into the notmuch database, and if they change we don't know how to hook it up properly?

If so I'm not sure we can fix this here. Some possible solutions (off the top of my head):
- require that if the maildir is under the notmuch dir, you have to move them together so the prefixes match
- find some other way to match message ids than by filename (maybe notmuch properties? otherwise hard, secondary index or something)
- force notmuch's `mail_root`to follow mujmap's `mail_dir` so we can use the global path query only (breaks anyone that has multiple maildirs in a single notmuch index)

Something else might be to do a sanity check. If there's a bunch of stuff in the index, but none of the paths exist on disk, abort and warn the user. Hmm, and thinking more on that, maybe the answer is that if a message is not on disk, but is in notmuch, and is on the server, re-download it.

I'm not really arguing against the need for `MailDirNotASubdirOfNotmuchRoot` as such, I just don't know when it should fire. But honestly, I still wonder if this is a thing. Is moving things around a common thing that notmuch people like to do? (I am only a new notmuch person myself). If its not, then maybe noticing something and off and forcing a resync is the right way.

(I wrote this over the course of a busy and distracted day so I feel like everything has fallen out of my head even more than the original plan fell out of my head in the last couple of months; I'm really struggling. I'll try to think about this more over the weekend).

Reply:
> So in your scenario, if they've set `mail_dir` to something `~/Maildir/mujmap` then they get what they ask for, regardless of whether or not its a subdir of the notmuch root. This has to be this way to support "side-by-side" configs (like XDG).

I think maybe this is where the confusion stems. notmuch doesn't actually support this configuration; everything *has* to be a subdirectory of `database.mail_root`. `database.path` used to refer to the same thing as `database.mail_root`, but is now independent and only refers to the actual database files if both of these options are specified ([see here](https://notmuchmail.org/doc/latest/man1/notmuch-config.html)). If you've somehow managed to index files outside of the root, I think that's actually undefined behavior and probably not good. :confused: But it seems like the best way to prevent unintended behavior is to catch this discrepancy between the two configurations as early as possible, hence `MailDirNotASubdirOfNotmuchRoot`.

My example was mostly just me trying to address what I thought your misconception was rather than your actual example, rather than me trying to necessarily specifically prevent these kinds of database migration issues, but I do agree that a sanity check for a really large unexpected change like this would be a great to have feature.

> (I wrote this over the course of a busy and distracted day so I feel like everything has fallen out of my head even more than the original plan fell out of my head in the last couple of months; I'm really struggling. I'll try to think about this more over the weekend).

That's okay, me too :slightly_smiling_face:

Reply:
> I think maybe this is where the confusion stems. notmuch doesn't actually support this configuration; everything has to be a subdirectory of database.mail_root. database.path used to refer to the same thing as database.mail_root, but is now independent and only refers to the actual database files if both of these options are specified ([see here](https://notmuchmail.org/doc/latest/man1/notmuch-config.html)). If you've somehow managed to index files outside of the root, I think that's actually undefined behavior and probably not good. confused

I agree that's where the confusion comes from; if using `NOTMUCH_PROFILE` these things can all be separated, and indeed, that's a large part of the reason I'm even doing this work!

My current notmuch config with profiles, fyi:

```
$ env NOTMUCH_PROFILE=home notmuch config list
...
database.backup_dir=/home/robn/.local/share/notmuch/home/backups
database.hook_dir=/home/robn/.local/share/notmuch/home/hooks
database.mail_root=/home/robn/.local/share/mujmap/home
database.path=/home/robn/.local/share/notmuch/home
```

And then I run mujmap with this PR as: `env NOTMUCH_PROFILE=home mujmap --path /home/robn/.config/mujmap/home.toml`. `mail_dir` is not set there, so its being taken from notmuch.

I did consider that maybe the intent of notmuch's profile support was still that the database was under the maildir, and maybe I was indeed getting accidental behaviour. The config doc is not clear on this, though perhaps that's just it not answering a question that no one asked. However, the [tests for split-path configs](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=test/T055-path-config.sh;h=fe2953247302db5d1a5da8568b5667c1a64a000c;hb=HEAD#l71) seems to imagine that the database path and mail dir can be in very different locations, as perhaps does the [commit that introduced it](https://nmbug.notmuchmail.org/nmweb/show/20210314140217.183425-15-david%40tethera.net), and a code read seems to support that too.

I think I don't have any more to add. I (of course) think I'm right about what notmuch is supposed to be able to do here; if you're not persuaded then I'm not sure what we do next.